### PR TITLE
Feature/noid/imagecaching

### LINKED ISSUE #7: Request: Pull image(s) in advance with cronjob to have it ready uppon loading the page
I experience a little delay when I load my cloud page until the image is shown. It would be great to pull the next image with the cronjob. Information if this image was then shown to the user could be kept in database. So that if a image was pre cached, it is no necessary to cache another one, until the previous was shown once.

Besides that ... I love your App (idea). It is simple and beautiful!

### LINKED ISSUE #9: URL to photo
I would love to know the url of the photo; display it in small print on the bottom, for example. It should probably be an option, as I'm sure there are many who wouldn't want that to show up on the fantastic login screen.

### LINKED ISSUE #123: Could not find resource unsplash/js/user.js to load + include(): Failed opening license.php #114
using Unraid + Docker (linuxserver)

v2.2.1 was working after updated + Nextcloud 27 .

but now not working again :  
1. Could not find resource unsplash/js/user.js to load
2. include(/app/www/public/apps/unsplash/templates/partials/license.php) https://github.com/nextcloud/unsplash/issues/114

image

### LINKED ISSUE #131: Splash - not supported on Nextcloud 28
use **[GitHub reactions](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/)** 👍 to show that you are affected by the same issue. Please don't comment if you have no relevant information to add!

- [ ] This is not a feature request
- [ ] This is not an image-provider request
- [ ] I have searched the issuetracker if the issue is not already reported.

**Describe the bug:**
I believe the 2.2.1 isn't supported on NextCloud 28

**To Reproduce:**
It's not displaying the images at the login as it normally would

**Expected behavior:**
Not showing on the login screen,  attempted to remove app and reinstall app,  rebooted services, server etc

**Screenshots:**
Screenshot will show the app successfully installed and no background photo

**Server:**
 - Nextcloud version:  28
 - Splash-App version: 2.2.1

**Additional Information:**
It's such a cool app,  hope to get it working again
Thanks
Daniel - Perth Western Australia.

### PR DESCRIPTION
**Pull request by newhinton**
- Adds image caching to some providers
- Support for nextcloud 28

closes #123 
closes #131
implements #115
implements #2 
closes #7 
closes #9

### CODE REVIEW

**Comment by EtienneHenger** [comment]
Is there an estimate as to when this will be merged into main? Asking as I am looking to update to NC28 and am wondering if it is worth the hassle to update this app via the terminal as opposed to waiting a bit and having it in the official app release...

**Comment by newhinton** [comment]
Generally speaking, this PR is feature-complete. However, since i wrote it over a couple of month, and it introduces some big changes, i want to do a full review before merging. That still might take a while. Also, any help is appreciated and will speed this up!

**Comment by EtienneHenger** [comment]
Thanks for the quick reply! I would be happy to help any way I can; so how can I help? Mind you I am a novice in PHP, JavaScript, and CSS...

File: lib/Provider/UnsplashAPI.php
Comment (mwinkens) [inline lib/Provider/UnsplashAPI.php:22]: This file (and a lot of other files) are full of alignment/linting issues, please run something of the lines of `php-cs-fixer`, at least for the new files. Maybe there needs to be a cleanup follow up
This makes reviewing even harder

File: lib/Provider/UnsplashAPI.php
Comment (mwinkens) [inline lib/Provider/UnsplashAPI.php:50]: commented out code

File: templates/partials/license.php
Comment (mwinkens) [inline templates/partials/license.php:1]: Do you need this file for anything as it's empty?

File: lib/Services/SettingsService.php
Comment (mwinkens) [inline lib/Services/SettingsService.php:187]: This doesn't return a string, but a `Provider`. It uses a string, maybe an Id would be better :)

File: lib/Services/SettingsService.php
Comment (mwinkens) [inline lib/Services/SettingsService.php:350]: idk, I would remove the @return here

File: lib/Provider/UnsplashAPI.php
Comment (mwinkens) [inline lib/Provider/UnsplashAPI.php:49]: why the double slash?

**Review by mwinkens — CHANGES_REQUESTED** [review]
Thank you for your PR! unsplash currently prevents me from updating to nc28, so this PR is **very welcomed**. I just request a few tiny changes, in order to give you some feedback and make this ready for merge. Feel free to challenge some of my review points, since I am not maintainer of this project.

Unfortunately I shouldn't review your javascript files and I still need to
- [ ] test this in production

**Comment by newhinton** [comment]
@mwinkens 

Thank you for your review! Your work is very appreciated!

I will try to work through your notes later, the linting-idea is a good one, and i'll do something about it!

**Comment by mwinkens** [comment]
> @mwinkens
> 
> Thank you for your review! Your work is very appreciated!
> 
> I will try to work through your notes later, the linting-idea is a good one, and i'll do something about it!

I also updated your description, feel free to change it accordingly! I noticed you mentioned the closed issues also in the commits

File: templates/partials/license.php
Comment (mwinkens) [reply templates/partials/license.php:1]: Ah this closes #114

File: lib/EventListener/BeforeTemplateRenderedEventListener.php
Comment (newhinton) [reply lib/EventListener/BeforeTemplateRenderedEventListener.php:74]: That is a good question. There probably was a reason (i assume it was to prevent breakage and provide a graceful fallback for "unknown" routes), but to be frank, i dont remember why i removed it in 2022. It took me a while to finish this PR :D

**Comment by newhinton** [comment]
I have not introduced an dedicated linter, however, i used my ide's tools to reformat the files. I am a bit hesitant to use a tool, because it seems most depend on some package manager like compose, and at the moment this app does not use one because it is rather simple, dependency-wise. 

But i have worked through your helpful annotations and a lot of that was helpful! Thank you for making that effort!

**Comment by mwinkens** [comment]
> Hello, thanks for the work, I tested and it works for the login page but not when you are logged in. How can I help you?

@newhinton I tested this PR now manually and I have the same problem. Normally you have a(n) (un)splash background when logged in, this is mostly visible in the dashboard. I can add screenshots if you want.

Otherwise this looks good, you still have a few conversations open (see above), I closed the ones that are resolved. Thank you for fixing the linting issues :+1:

File: lib/EventListener/BeforeTemplateRenderedEventListener.php
Comment (mwinkens) [reply lib/EventListener/BeforeTemplateRenderedEventListener.php:74]: I would reintroduce it. Some apps might have paths, that you can't know here. For example this might break public links

**Comment by mwinkens** [comment]
Also I managed to produce some exceptions with the UnsplashAPI provider (I didn't provide a valid token)

```
{"reqId":"3PdFVjay1aLCnS1Ix60x","level":3,"time":"2024-03-18T11:46:52+00:00","remoteAddr":"127.0.0.1","user":"nextcloud27","app":"index","method":"GET","url":"/index.php/apps/unsplash/api/dashboard.css","message":"/appdata_oc9arpx6pmoc/unsplash/UnsplashAPI","userAgent":"Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:123.0) Gecko/20100101 Firefox/123.0","version":"28.0.1.1","exception":{"Exception":"OCP\\Files\\NotFoundException","Message":"/appdata_oc9arpx6pmoc/unsplash/UnsplashAPI","Code":0,"Trace":[{"file":"/var/www/nextcloud28/lib/private/Files/Node/LazyFolder.php","line":161,"function":"get","class":"OC\\Files\\Node\\Root","type":"->"},{"file":"/var/www/nextcloud28/lib/private/Files/AppData/AppData.php","line":132,"function":"get","class":"OC\\Files\\Node\\LazyFolder","type":"->"},{"file":"/home/marvin/workspace/unsplash/lib/ProviderHandler/Provider.php","line":236,"function":"getFolder","class":"OC\\Files\\AppData\\AppData","type":"->"},{"file":"/home/marvin/workspace/unsplash/lib/Provider/UnsplashAPI.php","line":51,"function":"getImageFolder","class":"OCA\\Unsplash\\ProviderHandler\\Provider","type":"->"},{"file":"/home/marvin/workspace/unsplash/lib/Provider/UnsplashAPI.php","line":46,"function":"getMetadata","class":"OCA\\Unsplash\\Provider\\UnsplashAPI","type":"->"},{"file":"/home/marvin/workspace/unsplash/lib/Services/SettingsService.php","line":263,"function":"getCachedImageURL","class":"OCA\\Unsplash\\Provider\\UnsplashAPI","type":"->"},{"file":"/home/marvin/workspace/unsplash/lib/Controller/CssController.php","line":114,"function":"headerbackgroundLink","class":"OCA\\Unsplash\\Services\\SettingsService","type":"->"},{"file":"/home/marvin/workspace/unsplash/lib/Controller/CssController.php","line":88,"function":"innerCSS","class":"OCA\\Unsplash\\Controller\\CssController","type":"->"},{"file":"/home/marvin/workspace/unsplash/lib/Controller/CssController.php","line":58,"function":"mediaQuery","class":"OCA\\Unsplash\\Controller\\CssController","type":"->"},{"file":"/var/www/nextcloud28/lib/private/AppFramework/Http/Dispatcher.php","line":230,"function":"dashboard","class":"OCA\\Unsplash\\Controller\\CssController","type":"->"},{"file":"/var/www/nextcloud28/lib/private/AppFramework/Http/Dispatcher.php","line":137,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud28/lib/private/AppFramework/App.php","line":184,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud28/lib/private/Route/Router.php","line":315,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/var/www/nextcloud28/lib/base.php","line":1069,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/var/www/nextcloud28/index.php","line":39,"function":"handleRequest","class":"OC","type":"::"}],"File":"/var/www/nextcloud28/lib/private/Files/Node/Root.php","Line":207,"message":"/appdata_oc9arpx6pmoc/unsplash/UnsplashAPI","exception":{},"CustomMessage":"/appdata_oc9arpx6pmoc/unsplash/UnsplashAPI"}}
```

The other providers seem to work fine, but also seem to fill the cache folder, which is funny because the WikiCommon provider fills the nature photos with cars

**Comment by newhinton** [comment]
@mwinkens 

> I found out, that this is an admin setting, you can enable/disable splash for the login and dashboard.

So there is no issue at all and that part work for you now?

> Also I managed to produce some exceptions with the UnsplashAPI provider (I didn't provide a valid token)

Good catch!


> The other providers seem to work fine, but also seem to fill the cache folder

How bad is it, do you have some numbers?

**Comment by mwinkens** [comment]
> So there is no issue at all and that part work for you now?
No there isn't, this was just a configuration issue!
> The other providers seem to work fine, but also seem to fill the cache folder
The number of images isn't the problem, but with different providers the image content changes. If I switch between providers, I fill the image cache with cars and nature photos. Maybe the image cache needs to be deleted on provider switch?

Otherwise this looks good! I noticed, that there is a card on the dashboard and login screen with image information, but unfortunately it isn't clickable, is this part of this PR?

**Comment by newhinton** [comment]
> Otherwise this looks good! I noticed, that there is a card on the dashboard and login screen with image information, but unfortunately it isn't clickable, is this part of this PR?

It should be clickable, do you have javascript disabled? Could also be a cache problem, can you reload with caching disabled?

**Comment by JW-CH** [comment]
Love to see the progress on this, came back to this because v29 was released.

**Comment by mwinkens** [comment]
> > Otherwise this looks good! I noticed, that there is a card on the dashboard and login screen with image information, but unfortunately it isn't clickable, is this part of this PR?
> 
> It should be clickable, do you have javascript disabled? Could also be a cache problem, can you reload with caching disabled?

I disabled the cache and reloaded the page, but the problem persists. This button just redirects me from the dashboard to the dashboard

Bildschirmfoto vom 2024-04-29 09-42-34

Imo this PR is good enough, I would merge and address issues in follow ups!

**Comment by CharlesDelorme** [comment]
I would like to help ? How can I install this on my NC28 ?

**Review by mwinkens — APPROVED** [review]

**Comment by mwinkens** [comment]
> I would like to help ? How can I install this on my NC28 ?

 [clone PR](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally,) into nextcloud apps directory under the unsplash directory,  don't forget to change directory permission with `chown`, `php occ app:enable unsplash`

**Comment by CharlesDelorme** [comment]
> > I would like to help ? How can I install this on my NC28 ?
> 
> [clone PR](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally,) into nextcloud apps directory under the unsplash directory, don't forget to change directory permission with `chown`, `php occ app:enable unsplash`

Thank you ! I read the doc and I'll try that.

**Comment by asheroto** [comment]
When will the next release occur?

**Comment by newhinton** [comment]
I need to update the version and write a changelog, theb it'll be released for 28. I dont think suppoert for 29 requires changes, but i need to validate that.

**Comment by newhinton** [comment]
Oh, and i need to add some setup-documentation for the new unsplash api token.

**Comment by MiWCryptAnalytics** [comment]
Got it working on 29!

Maybe rename token to "access key" to match Unsplashed naming, i was using secret key and it failed.
It worked with the access key.

NB: code has Title case L in login - core.login.showLoginForm:
lib/EventListener/BeforeTemplateRenderedEventListener.php:63 should be case 'core.login.showloginform': 

This was causing error:

Call to undefined method OCP\\AppFramework\\Http\\Events\\BeforeLoginTemplateRenderedEvent::isLoggedIn()
It was hitting the default case due to the switch case spelling, and because a BeforeLoginTemplateRenderedEvent does not defne isLoggedIn.
App would HTTP/500 on the login screen. I tried handling this case specifically, but the root cause is the case spelling.

I might suggest not checking for isLoggedIn() for all cases, as BeforeLoginTemplateRenderedEvent does not provide this, and might fall through to default case for other BeforeLoginTemplateRenderedEvent event generating routes.

**Comment by CrazyWolf13** [comment]
Hi
Thanks for this PR.

What's the norma release flow of nextcloud? When can we expect an "official" update to the nextcloud store?

**Pull request merged**
